_posts/2018-05-23-getting-foreman-search-results-into-your-puppet-manifest.md

```diff
@@ -0,0 +1,98 @@
+
+But, if you don't have a PuppetDB and you are already using Foreman as external node classifier (aka ENC), you can also make fun stuff.
+
+## An use case out of standard usages.
+
+Use cases about monitoring or backups are standard. Lets try to build another example where we can use the Foreman search engine.
+
+Imagine that you are managing nodes with a Puppet Server and they are in `hostgroups` like `secure_base/role`. Where `secure_base` permits to apply all common settings to all your nodes in a secured environment. And `role` permits to apply settings dedicated to the role of a node. Now, lets say you have the constraint to not use a DNS server. And your customer would like to call hosts in a human way, so with names and not with IPs. I already met this use case. Fortunately, servers were using only one IP address.
+
+So you have to manage your `/etc/hosts` with a configuration management tool or you become crazy. 
+
+If each Puppet node execute the following code :
+
+```
+  $query = foreman({foreman_user => 'apiuser',
+                    foreman_pass => 'here_a_password',
+                    item         => 'hosts',
+                    search       => 'hostgroup_fullname ~ secure_base',
+  })
+
+  $nodes = $query['results']
+
+```
+
+In `$nodes`, you got all the needed informations about other nodes. You can check my statement by adding the following piece of code :
+
+
+```
+  $nodes.each | $_node | {
+    $_thiscert = $_node['certname']
+    $_thisip = $_node['ip']
+    notify {  "node ${_thiscert}" : message => "host ${_thisip}"}
+  }
+
+```
+
+Let's use `$nodes` in a template with a `file` resource to manage `/etc/hosts` and you got the trick.
+
+## Unknow function foreman()
+
+If you ran the code above, you probably got an error message like `Evaluation Error: Unknown function: 'foreman'`. By default, the function `foreman()` is not available.
+
+Several possibilities to get the function into your environement :
+
+ * You installed Foreman with `foreman-installer`. And so, you can simply copy `/usr/share/foreman-installer/modules/foreman/lib/puppet/parser/functions/foreman.rb` into your module managing `/etc/hosts`.
+ * Otherwise, you can [download the function foreman()](https://github.com/theforeman/puppet-foreman/blob/master/lib/puppet/parser/functions/foreman.rb) and use it into your module managing `/etc/hosts`.
+ * Probably a better way is to create a module named like `foreman-libs` containing the function `foreman()`. Add this module in your list of imported modules. To be imported by Foreman, remember that a module must contain a file `manifests/init.pp` and not contain errors (like those reported by `puppet parser validate <file.pp>`).
```

Comment:
Having an `init.pp` isn't required, because importing the function into Foreman isn't required. A module that just contains `lib` is totally fine for loading functions into Puppet.

Reply:
Note that creating a "libs" module is a really good idea, so I do think we want to keep this point - just remove the bit about importing it into Foreman, it's not needed :)

Reply:
Ah ?! Last time i tryed to import a module without a file `manifests/init.pp`, the module was not imported. I didn't checked this since a long time, but it was written somewhere in docs of Foreman.